File: linkgame-common/src/main/java/io/github/octcarp/linkgame/common/module/GridPos.java
package io.github.octcarp.linkgame.common.module;

import java.io.Serializable;

public record GridPos(int row, int col) implements Serializable {
    public GridPos {
        if (row < 0 || col < 0) {
            System.err.printf("Invalid grid position: (%d, %d)\n", row, col);
        }
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof GridPos pos) {
            return row == pos.row && col == pos.col;
        }
        return false;
    }
}

File: game-common/src/main/kotlin/io/github/octcarp/sustech/cs209a/linkgame/common/model/GridPos.kt
package io.github.octcarp.sustech.cs209a.linkgame.common.model

import java.io.Serializable

data class GridPos(
    val row: Int,
    val col: Int
) : Serializable {

    override fun equals(obj: Any?): Boolean {
        if (this === obj) {
            return true
        }
        if (obj == null || javaClass != obj.javaClass) {
            return false
        }
        val pos = obj as GridPos
        return row == pos.row && col == pos.col
    }

    init {
        if (row < 0 || col < 0) {
            System.err.printf("Invalid grid position: (%d, %d)\n", row, col)
        }
    }

    override fun hashCode(): Int {
        var result = row
        result = 31 * result + col
        return result
    }
}
